File: sidecar/remove_similar_images.py
import os
import subprocess
import sys
from pathlib import Path


def ensure_dependencies() -> None:
    try:
        import PIL  # noqa: F401
        import imagehash  # noqa: F401
        return
    except Exception:
        pass

    print("Installing required packages: Pillow imagehash")
    proc = subprocess.run(
        [sys.executable, "-m", "pip", "install", "Pillow", "imagehash"],
        capture_output=True,
        text=True,
    )
    if proc.stdout:
        print(proc.stdout.strip())
    if proc.stderr:
        print(proc.stderr.strip())
    if proc.returncode != 0:
        raise RuntimeError("Failed to install Pillow/imagehash")


def collect_images(folder: Path) -> list[Path]:
    exts = {".jpg", ".jpeg", ".png", ".webp", ".bmp"}
    return sorted([p for p in folder.iterdir() if p.is_file() and p.suffix.lower() in exts])


def group_similar(images: list[Path], threshold: int) -> list[list[Path]]:
    from PIL import Image, ImageFile
    import imagehash

    ImageFile.LOAD_TRUNCATED_IMAGES = True

    hashes: dict[Path, imagehash.ImageHash] = {}
    for image_path in images:
        try:
            with Image.open(image_path) as img:
                hashes[image_path] = imagehash.phash(img)
            print(f"OK hashed: {image_path.name}")
        except Exception as exc:
            print(f"ERR hash {image_path.name}: {exc}")

    groups: list[list[Path]] = []
    visited: set[Path] = set()
    keys = list(hashes.keys())
    for i, p1 in enumerate(keys):
        if p1 in visited:
            continue
        h1 = hashes[p1]
        group = [p1]
        for p2 in keys[i + 1 :]:
            if p2 in visited:
                continue
            if (h1 - hashes[p2]) <= threshold:
                group.append(p2)
                visited.add(p2)
        visited.add(p1)
        if len(group) > 1:
            groups.append(group)
    return groups


def remove_duplicates(groups: list[list[Path]]) -> list[Path]:
    removed: list[Path] = []
    for group in groups:
        keep = group[0]
        print(f"GROUP keep={keep.name} total={len(group)}")
        for dup in group[1:]:
            try:
                dup.unlink()
                removed.append(dup)
                print(f"REMOVED {dup.name}")
            except Exception as exc:
                print(f"ERR remove {dup.name}: {exc}")
    return removed


def main() -> int:
    if len(sys.argv) < 2:
        print("Usage: python remove_similar_images.py <folder_path> [--delete] [--threshold N]")
        return 1

    folder = Path(sys.argv[1]).expanduser().resolve()
    do_delete = "--delete" in sys.argv

    threshold = 5
    if "--threshold" in sys.argv:
        idx = sys.argv.index("--threshold")
        if idx + 1 < len(sys.argv):
            threshold = int(sys.argv[idx + 1])

    if not folder.exists() or not folder.is_dir():
        print(f"ERR folder not found: {folder}")
        return 1

    ensure_dependencies()
    images = collect_images(folder)
    print(f"IMAGES_FOUND={len(images)}")
    if not images:
        return 0

    groups = group_similar(images, threshold)
    print(f"SIMILAR_GROUPS={len(groups)}")

    if not groups:
        print("DEDUPE_DONE removed=0 remaining=" + str(len(images)))
        return 0

    if do_delete:
        removed = remove_duplicates(groups)
        remaining = len(collect_images(folder))
        print(f"DEDUPE_DONE removed={len(removed)} remaining={remaining}")
    else:
        total_to_remove = sum(max(0, len(g) - 1) for g in groups)
        print(f"PREVIEW_ONLY to_remove={total_to_remove}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: sidecar/remove_similar_images.mjs
#!/usr/bin/env node

import { createHash } from 'crypto';
import fs from 'fs';
import path from 'path';

const IMAGE_EXTENSIONS = new Set(['.jpg', '.jpeg', '.png', '.webp', '.bmp']);

function usage() {
    console.log('Usage: node remove_similar_images.mjs <folder_path> [--delete] [--threshold N]');
}

function collectImages(folderPath) {
    const entries = fs.readdirSync(folderPath, { withFileTypes: true });
    return entries
        .filter((entry) => entry.isFile())
        .map((entry) => entry.name)
        .filter((name) => IMAGE_EXTENSIONS.has(path.extname(name).toLowerCase()))
        .sort()
        .map((name) => path.join(folderPath, name));
}

function hashFile(filePath) {
    const data = fs.readFileSync(filePath);
    return createHash('sha256').update(data).digest('hex');
}

function groupByHash(imagePaths) {
    const groups = new Map();
    for (const imagePath of imagePaths) {
        try {
            const hash = hashFile(imagePath);
            console.log(`OK hashed: ${path.basename(imagePath)}`);
            const bucket = groups.get(hash) ?? [];
            bucket.push(imagePath);
            groups.set(hash, bucket);
        } catch (error) {
            console.log(`ERR hash ${path.basename(imagePath)}: ${error instanceof Error ? error.message : String(error)}`);
        }
    }
    return [...groups.values()].filter((group) => group.length > 1);
}

function removeDuplicates(groups) {
    let removed = 0;
    for (const group of groups) {
        const [keep, ...dups] = group;
        console.log(`GROUP keep=${path.basename(keep)} total=${group.length}`);
        for (const dup of dups) {
            try {
                fs.unlinkSync(dup);
                removed += 1;
                console.log(`REMOVED ${path.basename(dup)}`);
            } catch (error) {
                console.log(`ERR remove ${path.basename(dup)}: ${error instanceof Error ? error.message : String(error)}`);
            }
        }
    }
    return removed;
}

function main() {
    if (process.argv.length < 3) {
        usage();
        return 1;
    }

    const folder = path.resolve(process.argv[2]);
    const doDelete = process.argv.includes('--delete');

    // Kept for compatibility with previous interface.
    // This script currently performs exact-content dedupe only.
    const thresholdIdx = process.argv.indexOf('--threshold');
    if (thresholdIdx >= 0 && thresholdIdx + 1 < process.argv.length) {
        const thresholdValue = Number.parseInt(process.argv[thresholdIdx + 1], 10);
        if (Number.isFinite(thresholdValue)) {
            void thresholdValue;
        }
    }

    if (!fs.existsSync(folder) || !fs.statSync(folder).isDirectory()) {
        console.log(`ERR folder not found: ${folder}`);
        return 1;
    }

    const images = collectImages(folder);
    console.log(`IMAGES_FOUND=${images.length}`);
    if (images.length === 0) {
        return 0;
    }

    const groups = groupByHash(images);
    console.log(`SIMILAR_GROUPS=${groups.length}`);
    if (groups.length === 0) {
        console.log(`DEDUPE_DONE removed=0 remaining=${images.length}`);
        return 0;
    }

    if (!doDelete) {
        const toRemove = groups.reduce((sum, group) => sum + Math.max(0, group.length - 1), 0);
        console.log(`PREVIEW_ONLY to_remove=${toRemove}`);
        return 0;
    }

    const removed = removeDuplicates(groups);
    const remaining = collectImages(folder).length;
    console.log(`DEDUPE_DONE removed=${removed} remaining=${remaining}`);
    return 0;
}

process.exitCode = main();
